crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -243,8 +252,103 @@ fn validate_input_mesh(mesh: &Mesh) -> Result<Cow<'_, [u32]>, MeshToMeshletMeshC
+    indices: &[u32],
+    vertices: &VertexDataAdapter,
+    position_only_vertex_remap: &[u32],
+    position_only_vertex_count: usize,
+) -> Meshlets {
+    // For each vertex, build a list of all triangles that use it
+    let mut vertices_to_triangles = vec![Vec::new(); position_only_vertex_count];
+    for (i, index) in indices.iter().enumerate() {
+        let vertex_id = position_only_vertex_remap[*index as usize];
+        let vertex_to_triangles = &mut vertices_to_triangles[vertex_id as usize];
+        vertex_to_triangles.push(i / 3);
+    }
+
+    // For each triangle pair, count how many vertices they share
+    let mut triangle_pair_to_shared_vertex_count = <HashMap<_, _>>::default();
+    for vertex_triangle_ids in vertices_to_triangles {
+        for (triangle_id1, triangle_id2) in vertex_triangle_ids.into_iter().tuple_combinations() {
+            let count = triangle_pair_to_shared_vertex_count
+                .entry((
+                    triangle_id1.min(triangle_id2),
+                    triangle_id1.max(triangle_id2),
+                ))
+                .or_insert(0);
+            *count += 1;
+        }
+    }
+
+    // For each triangle, gather all other triangles that share at least one vertex along with their shared vertex count
+    let triangle_count = indices.len() / 3;
+    let mut connected_triangles_per_triangle = vec![Vec::new(); triangle_count];
+    for ((triangle_id1, triangle_id2), shared_vertex_count) in triangle_pair_to_shared_vertex_count
+    {
+        // We record both id1->id2 and id2->id1 as adjacency is symmetrical
+        connected_triangles_per_triangle[triangle_id1].push((triangle_id2, shared_vertex_count));
+        connected_triangles_per_triangle[triangle_id2].push((triangle_id1, shared_vertex_count));
+    }
+
+    // The order of triangles depends on hash traversal order; to produce deterministic results, sort them
+    for list in connected_triangles_per_triangle.iter_mut() {
+        list.sort_unstable();
+    }
+
+    let mut xadj = Vec::with_capacity(triangle_count + 1);
+    let mut adjncy = Vec::new();
+    let mut adjwgt = Vec::new();
+    for connected_triangles in connected_triangles_per_triangle {
+        xadj.push(adjncy.len() as i32);
+        for (connected_triangle_id, shared_vertex_count) in connected_triangles {
+            adjncy.push(connected_triangle_id as i32);
+            adjwgt.push(shared_vertex_count);
+            // TODO: Additional weight based on triangle center spatial proximity?
+        }
+    }
+    xadj.push(adjncy.len() as i32);
+
+    let mut options = [-1; metis::NOPTIONS];
+    options[metis::option::Seed::INDEX] = 17;
+    options[metis::option::UFactor::INDEX] = 1; // Important that there's very little imbalance between partitions
+
+    let mut meshlet_per_triangle = vec![0; triangle_count];
+    let partition_count = triangle_count.div_ceil(126); // Need to undershoot to prevent METIS from going over 128 triangles per meshlet
```

Comment:
by how much does it overshoot the partitions it makes if you set this to 128?

Reply:
1-2 triangles, which then leads to tiny meshlets with 1 triangle. Need to undershoot unfortunately :(